ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/SandboxCommandCompletionService.scala

```diff
@@ -53,106 +56,82 @@ class SandboxCommandCompletionService private (
       "Received request for completion subscription {}: {}",
       subscriptionId: Any,
       request)
-    val offsetOrError =
-      request.offset.fold[Future[Option[Ref.LedgerString]]](Future.successful(None))(
-        o =>
-          LedgerOffsetValidator
-            .validate(o, "offset")
-            .fold(
-              Future.failed, {
-                case domain.LedgerOffset.Absolute(value) => Future.successful(Some(value))
-                case domain.LedgerOffset.LedgerBegin => Future.successful(None)
-                case domain.LedgerOffset.LedgerEnd =>
-                  ledgerBackend.getCurrentLedgerEnd.map(Some.apply)
-              }
-          ))
-
-    Source
-      .fromFuture(offsetOrError)
-      .flatMapConcat(completionSourceWithOffset(request, _, subscriptionId))
+
+    val offsetOrError = for {
+      offset <- FieldValidations.requirePresence(request.offset, "offset")
+      convertedOffset <- LedgerOffsetValidator.validate(offset, "offset")
+    } yield convertedOffset
+
+    offsetOrError.fold(
+      Source.failed,
+      offset => completionSourceWithOffset(request, offset, subscriptionId))
   }
 
   private def completionSourceWithOffset(
       request: CompletionStreamRequest,
-      requestedOffset: Option[LedgerSyncOffset],
+      requestedOffset: domain.LedgerOffset,
       subscriptionId: String): Source[CompletionStreamResponse, NotUsed] = {
-    val requestingParties = request.parties.toSet
-    val requestedApplicationId = request.applicationId
-    ledgerBackend
-      .ledgerSyncEvents(requestedOffset)
-      .map { syncEvent =>
-        val checkpoint =
-          Some(
-            Checkpoint(
-              Some(fromInstant(syncEvent.recordTime)),
-              Some(LedgerOffset(LedgerOffset.Value.Absolute(syncEvent.offset)))))
-
-        syncEvent match {
-          case tx: LedgerSyncEvent.AcceptedTransaction
-              if isRequested(
-                requestedApplicationId,
-                requestingParties,
-                tx.applicationId,
-                tx.submitter) =>
+    //TODO: put these into a proper validator
```

Comment:
Can you add the validator in this PR? We already do party validation in other places and I think it'd be great to clean it up at once.

Reply:
alright, will do that.

Reply:
we agreed to do it as a follow-up PR instead, to unblock further work